tokio/src/net/tcp/stream.rs

```diff
@@ -640,11 +653,8 @@ impl TcpStream {
     ///
     /// [`read`]: fn@crate::io::AsyncReadExt::read
     /// [`AsyncReadExt`]: trait@crate::io::AsyncReadExt
-    pub async fn peek(&self, buf: &mut [u8]) -> io::Result<usize> {
-        self.io
-            .registration()
-            .async_io(Interest::READABLE, || self.io.peek(buf))
-            .await
+    pub async fn peek(&self, buf: &mut ReadBuf<'_>) -> io::Result<usize> {
+        poll_fn(|cx| self.poll_peek(cx, buf)).await
     }
```

Comment:
This should still go through the registration setup, because otherwise you can't have multiple concurrent calls to `peek`.